src/env.cc

```diff
@@ -298,7 +298,13 @@ bool Environment::EmitNapiWarning() {
 void Environment::EnvPromiseHook(v8::PromiseHookType type,
                                  v8::Local<v8::Promise> promise,
                                  v8::Local<v8::Value> parent) {
-  Environment* env = Environment::GetCurrent(promise->CreationContext());
+  auto context = promise->CreationContext();
+  auto dataIndex = node::Environment::kContextEmbedderDataIndex;
+  // If the context is undefined (not a node context) then skip.
+  if (context->GetEmbedderData(dataIndex)->IsUndefined()) {
```

Comment:
@bnoordhuis is this still undefined behavior?

Reply:
Hm. I’m reading up on the V8 code, and it seems like a real issue here is that this might actually perform an out-of-bounds read if the context has less than `dataIndex` fields…

I assume the check is left out more or less intentionally on the V8 side, so I guess the options here are:
- Add a method to `v8::Context` that tells us how many embedder data fields there are 
- Let Node attach some kind of property (e.g. a private symbol) on a context-specific object, like the global object or the `ExtrasBindingObject`, then use that to tell whether an Environment has been assigned to a context or not … this might be a bit slower but wouldn’t require changes to V8

/cc @nodejs/v8

Reply:
/cc @hashseed many a years ago I ran into similar trouble with the debug context and had similar thoughts to @addaleax. WDYT?

Reply:
how about `HasEmbedderData(index)`

Reply:
@xaviergonz Do you have any thoughts on this? In the end, it’s your PR, so you obviously have some say in what happens here.

In particular, wrt the second option, that should be doable without waiting for anything else to happen, and it would be nice if you could indicate whether you’d want to try it yourself or prefer for somebody else to take a stab.

(If you do: We’d be really, really glad if we can help in any way. Don’t be shy to ask questions, here or on [IRC](https://webchat.freenode.net/?channels=node-dev&uio=d4)!)

Reply:
@addaleax I'd prefer someone else to take a stab at it since TBH I'd not even know where to start...

Reply:
@xaviergonz That’s fine, I think one of us can do that in the next few days.

But if you’re interested, here’s how I imagine the approach would roughly look like:

- Add a new symbol to `PER_ISOLATE_PRIVATE_SYMBOL_PROPERTIES` in `env.h`
- In `AssignToContext` in `env-inl.h`, set that symbol as a property on `context->Global()`.
  - The value should be irrelevant, but using `env->as_external()` might make sense; one could use that while debugging to verify that everything works.
  - The relevant APIs for using those symbols are `GetPrivate()` and `SetPrivate()`; you can grep through our code base if you’re unsure how to use them
- At the location of this discussion, you could check whether `GetPrivate()` returns an `External` value or not

Reply:
Hmm, and how can I access the private symbol from EnvPromiseHook if it is stored on the environment and I have to use it to check if we can get such environment?

Reply:
```
  auto context = promise->CreationContext();
  auto promiseHookSymbol = context->Global()->GetPrivate(context, env->promise_hook_private_symbol); // i need an env to access the private symbol...
  if (!promiseHookSymbol->IsUndefined()) {
    return;
  }
  Environment* env = Environment::GetCurrent(context); // but the env is accessed here
```

Reply:
@xaviergonz Good point…

We could use `Private::ForApi` with a fixed string name, kind of like `Symbol.for()`, but we really should measure the overhead in that case…

Reply:
@addaleax From what I can see Private::ForApi returns a symbol that I'd need to store somewhere...

maybe not using private symbols instead? At context creation time
```
gobal.process._envEmbedded = true;
```
and then on that EnvPromiseHook checking for
```
if (global.process && global.process._envEmbedded) we are good to go
```

Reply:
i'd like to push for stopping putting random underscored things on process, we already have *plently* of stuff as it is.

Reply:
Another place could be the promise prototype and since you get the promise instance on that call :)

Reply:
@xaviergonz `Private::ForApi` would return the same symbol per isolate, if we pass in the same string, but yes, it’s not idea.

> maybe not using private symbols instead? At context creation time

Yes, that’s probably a better approach, and I agree with @devsnek that adding a visible property is probably not a good idea (not on the `process` object and not on other builtins…). However, we could::

- In `Environment::Start` add an internal field to the process object, and set it to some value we’d recognize
- Then here check whether `global.process` exists, is an object, has an internal field, and that internal field contains what we want it to contain

Reply:
> Hm. I’m reading up on the V8 code, and it seems like a real issue here is that this might actually perform an out-of-bounds read if the context has less than dataIndex fields…

If `V8_ENABLE_CHECKS` is not set, yes.

I'll admit that this API is not particularly well-designed. It fulfills its purpose, not more. You can either use `Context::SetAlignedPointerInEmbedderData` to store arbitrary aligned C++ pointers, or use `Context::SetEmbedderData` to store V8 values.

In this particular case, Node would expect a `Environment*` value in the `kContextEmbedderDataIndex` slot. If no such pointer has been stored, depending on how large the currently allocated backing store is, this could either lead to an OOB access, the `undefined` value, or an assertion failure if compiled with `V8_ENABLE_CHECKS`.

What I would actually prefer is to always store a value to the `kContextEmbedderDataIndex` for every new V8 context. In case of absense of a valid `Environment*`, store `nullptr`. This way, you would use `Context::GetAlignedPointerInEmbedderData` and check the return value against `nullptr`.

Maybe this is guarding against the case that the context comes from somewhere else than Node core? In that case I don't see a good solution, because you can't be sure the `kContextEmbedderDataIndex` is not used to store some arbitrary non-undefined data.

Reply:
Um... is there a reason we use `Object::CreationContext` to get the context? What it's supposed to guard, `Environment::GetCurrent`, takes the current context of the isolate. These may be two different contexts.

Besides, iirc `Object::CreationContext` is discouraged. We only use it in Chrome where we haven't gotten around to fix yet. To get the creation context of an object, you load the map, from there the constructor function, then the function context, and finally the native context.

I'd prefer simply adding a symbol property to the Promise prototype to check. Alternatively, add a `Context::GetNumberOfDataSlots()` similar to the already existing `Isolate::GetNumberOfDataSlots()` (though the latter simply returns a hard-coded constant), but checking for `IsUndefined()` still seems unsafe, if Chromium decides to use that slot for some value.

Reply:
@addaleax something like this on start?
```
  // Used by EnvPromiseHook to know that we are on a node context.
  process_object->SetInternalField(0, OneByteString(isolate_, "nodeContext"));
```

Reply:
@xaviergonz Yup, something like that. I think a “magic number” might be easier, creating strings does come with some noticeable overhead.

(I don’t know if it’s obvious to you, but you might need to call `->SetInternalFieldCount()` on the instance template for `process` first so that the external field exists to begin with)

Reply:
Hm... looking through the code, the current context at that point should be the same context as the one the Promise originates from, so that at least seems fine. We should then just use `Isolate::GetCurrentContext`. In fact, maybe we should move this safeguard into `Environment::GetCurrent`.

How about this:
- Use the embedder field indexed 0 and fill it with a constant magic value (e.g. a static address to avoid collision). This signals that we are looking at a node context. This is safe because Chromium uses at least one embedder field in the context.
- Add an API to get the number of allocated embedder fields so that we can make embedder field accesses safe in the future.

Reply:
Using the process object has the down side that you do an object lookup to get it from the global object. Sounds slow.

Reply:
@addaleax sadly process_template is a FunctionTemplate and not an object template, so it does not contain a SetInternalFieldCount...

Reply:
nevermind, found process_template->InstanceTemplate()->SetInternalFieldCount(1);

Reply:
Fwiw, @hashseed’s suggestion sounds fine to me too.

Reply:
That would not be hogging 0 index for Node. You would check whether the value at index 0 is the magic number. If yes, we have a Node context. If not, we don't, and you leave the context alone, regardless of how it is used. Only requirement is that it is used so that we don't read OOB.

Reply:
This is what I did in the end that seems to work inside electron:
For env start:
```
  auto process_template = FunctionTemplate::New(isolate());
  process_template->SetClassName(FIXED_ONE_BYTE_STRING(isolate(), "process"));
  process_template->InstanceTemplate()->SetInternalFieldCount(1);

  auto process_object =
      process_template->GetFunction()->NewInstance(context()).ToLocalChecked();
  set_process_object(process_object);

  SetupProcessObject(this, argc, argv, exec_argc, exec_argv);

  // Used by EnvPromiseHook to know that we are on a node context.
  process_object->SetInternalField(0, v8::Int32::New(isolate(), 0x6e6f6465));
```

for EnvPromiseHook:
```
  auto context = promise->CreationContext();
  auto isolate = context->GetIsolate();
  auto global = context->Global();

  // Make sure process is there and its first internal field is the magic value.
  auto process = global->Get(OneByteString(isolate, "process"));
  if (!process->IsObject()) {
    return;
  }
  auto process_object = process->ToObject();
  if (process_object->InternalFieldCount() < 1) {
    return;
  }
  auto internal_field = process_object->GetInternalField(0);
  if (!internal_field->IsInt32()) {
    return;
  }
  auto internal_field_string = internal_field->ToInt32();
  if (!internal_field_string->StrictEquals(v8::Int32::New(isolate, 0x6e6f6465))) {
    return;
  }

  Environment* env = Environment::GetCurrent(context);
  for (const PromiseHookCallback& hook : env->promise_hooks_) {
    hook.cb_(type, promise, parent, hook.arg_);
  }
```

Should I change the pull request then?

Reply:
@hashseed It would still keep embedders from using that fieldfor Node Contexts – I guess that’s the original reason why we chose 32 as the index for the `Environment*` pointer…

Reply:
@xaviergonz I would change it to that, yes.

If you want review comments upfront: `ToInt32()` without a `context` argument is deprecated, but it’s unnecessary anyway; if you know that `IsInt32()` returned true, you can just cast & compare `internal_field.As<Int32>()->Value() == 0x6e6f6465`.

(`ToInt32()` would be similar to `a | 0` in JS – it actually performs a conversion of `a` if it’s not already an int32.)

Also, it might be good to write out the actual types instead of using `auto` – that helps with readability quite a bit, imo.

Apart from those nits, this looks good to me!

Reply:
@xaviergonz I'm worried that getting property lookup is not negligible. This will not make promise hooks any faster. Also, please do not use `CreationContext` for the reason I mentioned above.

@addaleax Oh...

Reply:
Btw I would be a bit surprised if index 0 was used by anybody. V8 used to use it for the [inspector](https://cs.chromium.org/chromium/src/v8/include/v8.h?l=8600&rcl=598a80edc4157838591a393a728a2f4f79a3b66f) until maybe a year ago.

Reply:
@hashseed IIRC `CreationContext` was literally the answer I got when I asked (I think it was gsathya) for how we should work around the fact that `PromiseHook`s are missing an opaque pointer that is passed through the API…

I’m not thrilled about that either.

Reply:
I'm somewhat certain that `Isolate::GetCurrentContext` should give you the same context, since we run the microtask in the context of the Promise. If `Object::CreationContext` and `Isolate::GetCurrentContext` returns different contexts, you got a problem anyways, since `Environment::GetCurrent` uses the latter.

Reply:
@hashseed How do we get the isolate without `Isolate::GetCurrent()` or `promise->GetIsolate()`? Both seem worse than what we have now…

Reply:
`promise->GetIsolate` get's it by masking off the object address to get to the heap page, where the isolate address is stored. That's not too bad.
`Isolate::GetCurrent` reads it from TLS, which can be slow.

Reply:
> `promise->GetIsolate` get's it by masking off the object address to get to the heap page, where the isolate address is stored. That's not too bad.

> V8_DEPRECATE_SOON("Keep track of isolate correctly", Isolate* GetIsolate());

:cry:

Reply:
Yeah [about that](https://bugs.chromium.org/p/v8/issues/detail?id=7288)...